src/main/java/com/github/manolo8/darkbot/backpage/DispatchManager.java

```diff
@@ -0,0 +1,172 @@
+            } while (m.find());
+
+            return m.groupCount() == consumers.size();
+        }
+
+        public static boolean updateAll(String page, DispatchData data) {
+            boolean updated = true;
+            for (InfoReader reader : InfoReader.values()) {
+                updated &= reader.update(page, data);
+            }
+            return updated;
+        }
+    }
+
+    public boolean hireRetriever(Retriever retriever){
+        if(data.getAvailableSlots() > 0) {
+            //retriever.getCost(); // TODO: check cost
+            try {
+                String response = main.backpage.getConnection("ajax/dispatch.php", Method.POST)
+                        .setRawParam("command", "sendDispatch")
+                        .setRawParam("dispatchId", retriever.getId())
+                        .getContent();
+                if(response.contains("\"result\":\"ERROR\"")){
+                    System.out.println("Failed to dispatch " + retriever.getId() + ": " + response);
+                    update(10_000);
+                }else{
+                    System.out.println("Successfully dispatched " + retriever.getId() + ": " + response);
+                    update(10_000);
+                    return true;
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+                System.out.println("Exception hiring dispatcher: " + e);
+            }
+        }
+        return false;
+    }
+
+    public boolean collect(InProgress progress) {
+        try {
+            if(progress.getCollectable().equals("0")) return false;
+            System.out.println("Collecting: Slot " + progress.getSlotID());
+            String x = main.backpage.getConnection("ajax/dispatch.php", Method.POST)
+                    .setRawParam("command", "collectDispatch")
+                    .setRawParam("slot", progress.getSlotID())
+                    .getContent();
+
+            //parse response
+            //{"result":"OK","message":"Collected the following:","rewardsLog":[{"lootId":"Solidus","amount":3},{"lootId":"PLT-2026","amount":97},{"lootId":"Scrap","amount":3}]}
+
+            if (x.contains("\"result\":\"ERROR\"")) {
+                System.out.println("Unable to collect retriever");
+            } else {
+                System.out.println("Dispatch Collected: " + progress.getSlotID() + " : " + x.substring(x.indexOf("rewardsLog")));
+                return true;
+            }
+            update(1000);
+            //remove or do something with slot / empty?
+        } catch (Exception e) {
+            e.printStackTrace();
+            System.out.println("Exception collecting dispatcher: "+ e.toString());
+        }
+        return false;
+    }
+
+    public List<String> collectAll() {
+        return data.getInProgress().values().stream().filter(this::collect).filter(Objects::nonNull).map(InProgress::getName).collect(Collectors.toList());
```

Comment:
```suggestion
        return data.getInProgress().values().stream()
                .filter(Objects::nonNull)
                .filter(this::collect)
                .map(InProgress::getName)
                .collect(Collectors.toList());
```
You were relying on the catch exception to catch the NPE of null progress????
NPE is not an exception you should handle by catching, you should not let it happen to start with

also, newlines help reading this stream way easier

Reply:
the null was before when it was returning back a string with the name, forgot to remove it